### PR TITLE
feat(device.lua): add custom options for tolino

### LINKED ISSUE #9175: FR: Tolino Vision 5 buttons dont invert based on rotation
**Does your feature request involve difficulty completing a task? Please describe.**
Currently the buttons on a `Tolino Vision 5` do not invert based on document view rotation

**Describe the solution you'd like**
Option to enable invert of tolino's buttons (based on rotation and maybe manually)

**Describe alternatives you've considered**
From what i can tell, there is also currently no button to invert the buttons manually (at least i have not found a setting)

**Additional context**
I am still new to koreader and i use it now on my Tolino Vision 5 (Android 4.4), where koreader is installed through F-Droid, koreader version is `2022.05.1`

I have also already tried the options `Tap and Gestures > Page turns > Invert page turn buttons`, which did not seem to affect the Button page turn

~~Also i have noticed that the manual Rotation is only for the Document Viewer and when going back to the file browser it de-rotates again, is this intentional?~~ found setting "Keep orientation across Views"

PS: is it currently possible to have automatic rotation (preferably asked before rotation) based on the gyroscope? (if available)

Update: from what i can tell, the `Tolino Vision 5` is more or less the same as `Kobo Libra H2O`

### PR DESCRIPTION
**Pull request by hasezoey**
This PR adds a `device/tolino` directory with `device.lua` and `event_map.lua` and for now only detects and uses custom event_map configuration for model `E70K00`, uses android as base and fallback if model is not detected

fixes #9175

requires https://github.com/koreader/android-luajit-launcher/pull/381 (this is why this PR is a DRAFT)

### CODE REVIEW

**Comment by NiLuJe** [comment]
I'd possibly reduce and clarify the mapping by just loading the android table in... a table; and just overriding the few fields you need in that, and return *that*.

It'll make the file smaller, and the intent much clearer ;).

File: frontend/device/tolino/device.lua [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+-- Tolino Vision 5
+local TolinoVision5 = TolinoDevice:new{
+    model = "Tolino_vision_5",
+    hasEinkScreen = yes
+}
+
+if android.prop.hardwareType == "E70K00" then
+    return TolinoVision5
+else
+    logger.warn("unrecognized Tolino model ".. android.prop.product.hardwareType.. " using android generic")
```
Comment (NiLuJe) [inline frontend/device/tolino/device.lua:26]: nit: `generic android` ;).

**Comment by hasezoey** [comment]
> I'd possibly reduce and clarify the mapping by just loading the android table in... a table; and just overriding the few fields you need in there, and return _that_.
> 
> It'll make the file smaller, and the intent much clearer ;).

i am still quite new to lua (and java for that matter), so i dont quite know to consolidate what to where
if you mean putting the changed buttons into `device/android/device_map`, then it is not really possible without overwriting the existing mappings for these values (values used are `21` and `22`)

**Comment by NiLuJe** [comment]
I simply meant that your new event map should essentially be

```lua
local keymap = require("blah/blah/android/event_map")
keymap.ChangedKey = "New_Event"

return keymap
```

(Obviously, fix the path first, I don't remember 'em by rote ;p).

File: frontend/device/tolino/device.lua (hunk 1 above)
Comment (hasezoey) [reply frontend/device/tolino/device.lua]: fixed with 6e46acd94d3e63f5012809f6111b9294c68b7335

**Comment by hasezoey** [comment]
> I simply meant that your new event map should essentially be something to the effect of

fixed with b7df2ee136e06a6a4676a06af5855b845108dd3e

**Comment by hasezoey** [comment]
in cf207b2d663b8b2f1e09a1d22a6d0fe465a11714 i changed the name from `TolinoVision5` (including property `model`) to rather be based on `ro.product.hardwareType` (or `Build.HARDWARE`) because there is no actual indication in android itself that it is a `Vision 5`

**Comment by hasezoey** [comment]
with c7bf69cc789bb8555a828e42c97ff4e93f0701b2 i added property `frontlight_settings` for `TolinoE70K00` because those leds exists and i could figure out what is what (thanks to the base tolino app), but i think it will not have any effect because those files are not writeable

for context that is what the following values are at `25`% color temp (what the tolino app reports as 25 in logcat):
```sh
$ cat /sys/class/backlight/lm3630a_led/brightness
15
$ cat /sys/class/backlight/lm3630a_leda/brightness
133
$ cat /sys/class/backlight/lm3630a_ledb/brightness
87
```

also assuming it is RGB because of
```sh
$ cat /sys/class/backlight/lm3630a_led/max_brightness
255
```
and that the color in a color picker looks similar enough to the color displayed on device (for quick reference `#85570f` is the hex color)

<details>
<summary>For Reference a Color Picker with the values put in</summary>

color_picker

</details>

**Comment by NiLuJe** [comment]
The lights stuff is indeed going to be useless without root, but, for curiosity's sake, take a look at how this is handled on Kobo, because that's the same controller (although I'm pretty sure we use the kernel's tables for those instead of setting each led individually).

**Comment by hasezoey** [comment]
> The lights stuff is indeed going to be useless without root, but, for curiosity's sake, take a look at how this is handled on Kobo, because that's the same controller (although I'm pretty sure we use the kernel's tables for those instead of setting each led individually).

from what i can tell in `Onxy*Controller.kt` files in `/sys/class/backlight/`(`white`/`warm`) are used to set the warmth, which dont exist in tolino
and in `device/kobo/device.lua` `frontlight_settings` are also set with leds(sysfs), which in the tolino are not writeable
and `device/kobo/powerd.lua` uses the nickel_conf, which has a path of `/mnt/onboard/.kobo/Kobo/Kobo eReader.conf`, but does not exist in tolino
and from what i can tell, kobo does not use `ntx_io` to set any light stuff, though that thing exists in tolino, but i dont know what to poke to try if light would work

**Comment by hasezoey** [comment]
Update to my previous search for setting the warmth on Tolino, ref: https://github.com/koreader/android-luajit-launcher/issues/351#issuecomment-1242981985

TL;DR: it is possible with `/dev/ntx_io` id `248` (scale 0-10)

**Review by pazos — CHANGES_REQUESTED** [review]
Please:

1. Do not touch `frontend/device.lua`
2. Do not add `frontend/device/tolino` files

The only change I see is in a couple of key remaps. If you want to do that just do it in `frontend/android/device.lua`.

Something along the lines of:

```lua
-- generic android event map
local android_event_map = require("device/android/event_map")

-- my tolino does that
if condition then
    android_event_map[21] = "LPgBack"
    android_event_map[22] = "LPgFwd"
end

self.input.event_map = android.event_map
```

please use `android.product` to check your condition.

File: frontend/device/tolino/device.lua [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+function TolinoDevice:init()
+    AndroidDevice.init(self)
+    self.input.event_map = require("device/tolino/event_map")
+end
+
+-- Tolino Vision 5, hardware model "E70K00"
+local TolinoE70K00 = TolinoDevice:new{
+    model = "Tolino_E70K00",
+    hasEinkScreen = yes,
```
Comment (pazos) [inline frontend/device/tolino/device.lua:20]: I'm afraid this doesn't work that way.

Here you're saying: hey I'm a eink screen. Let me refresh myself, but you didn't specify the driver kind, the type of waveforms you support....

With that option alone you won't get proper refreshes. Only a handful of messages on the log buffer saying that FakeEPDController issued some updates.

File: frontend/device/tolino/device.lua [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    frontlight_settings = {
```
Comment (pazos) [inline frontend/device/tolino/device.lua:21]: even if that works, the proper abstraction is in android-luajit-launcher, for both epd and lights driver. Please do that there so you won't break other widgets (ie: the native frontlight widget doesn't understand your code here)

Comment (hasezoey) [reply frontend/device/tolino/device.lua:21]: > even if that works, the proper abstraction is in android-luajit-launcher

i basically copied it from `device/kobo/device` for future reference

**Comment by hasezoey** [comment]
> please use `android.product` to check your condition.

i dont know if this would be correct, because that would report `ntx_6sl`,which would also match for `Tolino Shine 3` but having a different hardwareType (Vision5 is `E70K00` where as Shine3 is `e60k00`) - i dont know if the other models use the same mapping

**Comment by pazos** [comment]
> > please use `android.product` to check your condition.
> 
> i dont know if this would be correct, because that would report `ntx_6sl`,which would also match for `Tolino Shine 3` but having a different hardwareType (Vision5 is `E70K00` where as Shine3 is `e60k00`) - i dont know if the other models use the same mapping

They're using the same mapping.

IMHO this PR can be replaced with a simple user keymapping, as already covered in the wiki.

https://github.com/koreader/koreader/wiki/Android-tips-and-tricks#example-2-add-support-for-the-custom-keys-used-on-the-tolino-epos-2

If you want to workaround this without the need for user specific keymappings then please add an `IS_TOLINO` in android-luajit-launcher's DeviceInfo that includes current `TOLINO` definition and your own `TOLINO_VISION5` (in case your vision isn't recognized already as a `TOLINO`) and just do a generic check here.

**Comment by hasezoey** [comment]
> If you want to workaround this without the need for user specific keymappings then please add an IS_TOLINO in android-luajit-launcher's DeviceInfo that includes current TOLINO definition and your own TOLINO_VISION5 (in case your vision isn't recognized already as a TOLINO) and just do a generic check here.

from what i can tell all tolinos have `android.prop.model` as `tolino`, i am just unsure to add it for all tolinos not knowing if they use the same buttons and events

---

also thanks for linking me to the wiki, i kinda forgot github has this because other projects either dont have it or are just template or are way too outdated

**Comment by hasezoey** [comment]
@pazos is #9509 better? when yes, i will close this PR

**Comment by hasezoey** [comment]
closing in favor of https://github.com/koreader/koreader/pull/9509 and https://github.com/koreader/android-luajit-launcher/pull/382
